### ISSUE TITLE
Add support for function-like macros

### ISSUE THREAD

**Issue by gsingh93** [bpftrace/bpftrace#2459]
https://github.com/iovisor/bpftrace/pull/540 added support for [object-like](https://gcc.gnu.org/onlinedocs/cpp/Object-like-Macros.html#Object-like-Macros) macros, but it would also be nice to have support for [function-like](https://gcc.gnu.org/onlinedocs/cpp/Function-like-Macros.html#Function-like-Macros) macros.

I have some code in a probe that is a bit verbose, and I need to use almost the same code but with one minor change in another place. Since you can't define your own functions (as far as I know), being able to define function-like macros seems like it would be the best solution for this case.

One thing that might complicate this is to have useful function-like macros, we probably need to support multiline macros, and I don't think even object-like macros have multiline support.

I could just switch to `bcc` to use custom functions, but `bpftrace` supports everything I need to do, it's just I'd like the code for it to be more readable/maintainable.

**Comment by fbs** [comment]
personally I'd like to see something built into the language instead of a side effect of using clang for c header parsing. c macros are quite horrible

**Comment by gsingh93** [comment]
I agree! I just figured macros might be easier to implement given there's already support for a subset of them. But a proper language construct would be great.

**Comment by viktormalik** [comment]
I'd suspect function-like macros to be much more complicated to handle than simple one-line object-like macros that we support now. I'd rather invest energy into implementing proper function support, though that would probably require even more work.

In the meantime, you can manually preprocess the `.bt` file with a C preprocessor which will allow you to use full strength of C macros.

**Comment by ajor** [comment]
I'm fully behind us adding support for real functions in bpftrace.

That being said, I can still see a use for function-like macros. For example, we'd be able to implement `container_of` (#1455)  with the existing set of builtins.

**Comment by danobi** [comment]
Closing in favor of https://github.com/bpftrace/bpftrace/issues/3775
